contracts/vault/src/execute.rs

```diff
@@ -0,0 +1,591 @@
+    let config = state::CONFIG.load(deps.storage)?; // Load configuration
+    let sender = info.sender.clone(); // Address of the user requesting withdrawal
+
+    // Create a message to burn the user's USDCLP tokens
+    let burn_msg = WasmMsg::Execute {
+        contract_addr: config.usdclp_address.to_string(),
+        msg: to_json_binary(&Cw20ExecuteMsg::BurnFrom {
+            owner: sender.to_string(),
+            amount,
+        })?,
+        funds: vec![],
+    };
+
+    // Update the pending withdrawal amount and reduce the LP supply
+    state::PENDING_WITHDRAWALS.update(
+        deps.storage,
+        sender.as_str(),
+        |p| -> Result<Uint128, StdError> { Ok(p.unwrap_or(Uint128::zero()) + amount) },
+    )?;
+    state::TOTAL_LP_SUPPLY.update(deps.storage, |t| -> Result<Uint128, StdError> {
+        Ok(t - amount)
+    })?;
+
+    // Return a response with the burn message and attributes
+    Ok(Response::new().add_message(burn_msg).add_attributes(vec![
+        ("action", "request_withdrawal"),
+        ("user", sender.as_str()),
+        ("amount", &amount.to_string()),
+    ]))
+}
+
+/// Redistributes excess funds to markets based on their utilization
+///
+/// # Parameters
+/// - `deps`: Mutable dependencies
+/// - `env`: Contract environment
+/// - `info`: Message information
+///
+/// # Returns
+/// - `StdResult<Response>`: Response with deposit messages and attributes
+fn execute_redistribute_funds(deps: DepsMut, env: Env, info: MessageInfo) -> StdResult<Response> {
+    check_not_paused(&deps.as_ref())?;
+    if !is_authorized(&deps.as_ref(), &info.sender)? {
+        return Err(StdError::generic_err("Unauthorized"));
+    }
+
+    let config = state::CONFIG.load(deps.storage)?;
+    let pending: Uint128 = state::PENDING_WITHDRAWALS
+        .range(deps.storage, None, None, Order::Ascending)
+        .try_fold(Uint128::zero(), |acc, res| -> Result<Uint128, StdError> {
+            Ok(acc + res?.1)
+        })?;
+    let vault_balance = deps
+        .querier
+        .query_balance(&env.contract.address, &config.usdc_denom)?
+        .amount;
+    let excess = vault_balance
+        .checked_sub(pending)
+        .unwrap_or(Uint128::zero());
+
+    if excess.is_zero() {
+        return Err(StdError::generic_err("No excess to redistribute"));
+    }
+
+    // Retrieve the list of markets from the factory
+    let mut markets = vec![];
+    let mut start_after: Option<String> = None;
+    loop {
+        let batch: GetMarketsResponse =
+            deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+                contract_addr: config.factory_address.to_string(),
+                msg: to_json_binary(&FactoryQueryMsg::GetMarkets {
+                    start_after: start_after.clone(),
+                    limit: Some(30),
+                })?,
+            }))?;
+        if batch.markets.is_empty() {
+            break;
+        }
+        markets.extend(batch.markets);
```

Comment:
This includes all markets, even those that don't use USDC as the collateral asset.